cli/src/main.cpp

```diff
@@ -142,11 +153,101 @@ int main(int argc, char *argv[])
 		downloader->getUrls();
 	else if (parser.isSet(downloadOption))
 		downloader->getImages();
+	else if (parser.isSet(jsonOption)) {
+		downloader->setQuit(false);
+		downloader->getImages();
+		QObject::connect(downloader, &Downloader::finishedImages, [&](const QList<QSharedPointer<Image>> &images){
+			if (parser.isSet(loadDetailsOption))
+				loadMoreDetails(images);
+
+			QJsonDocument jsonDoc;
+			QJsonArray jsonArray;
+
+			for (auto& image : images) {
+				auto tokens = image->tokens(profile);
+				auto jsObject = serializeImg(image.data(), tokens);
+				jsonArray.append(jsObject);
+			}
+
+			jsonDoc.setArray(jsonArray);
+			auto jsonResult = jsonDoc.toJson(QJsonDocument::Indented);
+			QTextStream(stdout) << qPrintable(jsonResult);
+			app.quit();
+		});
+	}
 	else
 		parser.showHelp();
 
-	downloader->setQuit(true);
 	QObject::connect(downloader, &Downloader::quit, qApp, &QCoreApplication::quit);
 
 	return app.exec();
 }
+
+void loadMoreDetails(const QList<QSharedPointer<Image>> &images)
+{
+	int work = images.length();
+	QEventLoop loop;
+	int requestsLimit = 5;  // simultan requests
+	int runningRequests = 0;
+	for (auto& image : images) {
+		while (runningRequests >= requestsLimit)
+			QCoreApplication::processEvents(QEventLoop::AllEvents, 100);
+		runningRequests ++;
+		image->loadDetails();
+		QObject::connect(image.data(), &Image::finishedLoadingTags, [&](){
+			work--;
+			runningRequests--;
+			if (!work) loop.quit();
+		});
+	}
+	loop.exec();
+}
+
+QJsonObject serializeImg(const Image *image, const QMap<QString, Token> &tokens) {
+	static QStringList ignoreKeys = {"all", "allo", "allos", "all_namespaces", };
+	QJsonObject jsObject;
+
+	for (auto& key : tokens.keys()) {
+		typedef QVariant::Type Type;
+		if (ignoreKeys.contains(key)) continue;
+		if (key.contains("search_")) continue;
+
+		const QVariant& qvalue = tokens.value(key).value();
+		auto type = qvalue.type();
+
+		if (type == QVariant::Type::StringList) {
+			QStringList l = qvalue.toStringList();
+			if (l.isEmpty()) continue;
+			jsObject.insert(key, QJsonArray::fromStringList(l));
+		}
+		else if (type == QVariant::Type::String) {
+			QString s = qvalue.toString();
+			if (s.isEmpty()) continue;
+			jsObject.insert(key, s);
+		}
+		else if (type == Type::Url || type == Type::ULongLong || type == Type::LongLong)
+			jsObject.insert(key, qvalue.toString());
+		else if (type == Type::Int)
+			jsObject.insert(key, qvalue.value<int>());
+		else if (type == Type::Bool)
+			jsObject.insert(key, qvalue.value<bool>());
+		else if (type == Type::DateTime)
+			jsObject.insert(key, static_cast<int>(qvalue.value<QDateTime>().toTime_t()));
+		else {
+			qDebug() << qvalue;
+			log(QStringLiteral("using generic QVariant::toString for key: %1").arg(key), Logger::Warning);
```

Comment:
Does this currently ever happen, or it it only a fail-safe? (which is good 👍)

Reply:
Not that I know